analyzeme/src/lightweight_event.rs

```diff
@@ -0,0 +1,64 @@
+use crate::event::Event;
+use crate::profiling_data::ProfilingData;
+use crate::timestamp::Timestamp;
+use std::hash::{Hash, Hasher};
+use std::time::Duration;
+
+#[derive(Clone, Debug)]
+pub struct LightweightEvent<'a> {
+    pub data: &'a ProfilingData,
+    pub event_index: usize,
+    pub timestamp: Timestamp,
+    pub thread_id: u32,
+}
+
+impl<'a> LightweightEvent<'a> {
+    pub fn to_event(&self) -> Event<'a> {
+      self.data.decode_full_event(self.event_index)
+    }
+
+    /// Returns true if the time interval of `self` completely contains the
+    /// time interval of `other`.
+    pub fn contains(&self, other: &LightweightEvent) -> bool {
+        match self.timestamp {
+            Timestamp::Interval {
+                start: self_start,
+                end: self_end,
+            } => match other.timestamp {
+                Timestamp::Interval {
+                    start: other_start,
+                    end: other_end,
+                } => self_start <= other_start && other_end <= self_end,
+                Timestamp::Instant(other_t) => self_start <= other_t && other_t <= self_end,
+            },
+            Timestamp::Instant(_) => false,
+        }
+    }
+
+    pub fn duration(&self) -> Option<Duration> {
+        match self.timestamp {
+            Timestamp::Interval { start, end } => end.duration_since(start).ok(),
+            Timestamp::Instant(_) => None,
+        }
+    }
+}
+
+impl<'a> PartialEq for LightweightEvent<'a> {
+    fn eq(&self, other: &LightweightEvent<'a>) -> bool {
+        std::ptr::eq(self.data, other.data) &&
+            self.event_index == other.event_index &&
+            self.timestamp == other.timestamp &&
+            self.thread_id == other.thread_id
+    }
+}
+
+impl<'a> Eq for LightweightEvent<'a> {}
+
+impl<'a> Hash for LightweightEvent<'a> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        std::ptr::hash(self.data, state);
+        self.event_index.hash(state);
+        self.timestamp.hash(state);
+        self.thread_id.hash(state);
```

Comment:
While doing lots of `HashStable` impls in the compiler I've really come to appreciate using this "exhaustive destructuring" pattern in order not to forget any fields (and guard against forgetting newly added fields): 
```rust
let LightweightEvent {
    data,
    event_index,
    timestamp,
    thread_id,
} = *self;

std::ptr::hash(data, state);
event_index.hash(state);
timestamp.hash(state);
thread_id.hash(state);
```

It's a bit more verbose but it enables the compiler to help you with a common error case.

Reply:
This comment also applies to the `PartialEq` impl above.